paddle/fluid/operators/filter_instag_op.h

```diff
@@ -0,0 +1,160 @@
+#else
+template <typename T>
+using Vector = framework::CPUVector<T>;
+#endif
+
+template <typename T>
+class FilterInstagKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    // X1 is global FC output
+    // Dim [batch size, embedding size]
+    auto* x1 = context.Input<LoDTensor>("X1");
+    // X2 is ins tag list
+    // LoD [[0, Sum(ins1), Sum(ins1, ins2), ... ]]
+    auto* x2 = context.Input<LoDTensor>("X2");
+    // X3 is local fc tag list
+    // LoD [[0, Sum(fc1), Sum(fc1, fc2) ...]]
+    auto* x3 = context.Input<Tensor>("X3");
+
+    std::unordered_set<int64_t> filter_tag;
+    auto* x3_data = x3->data<int64_t>();
+    size_t len = x3->dims()[0];
+    for (size_t i = 0; i < len; i++) {
+      filter_tag.insert(x3_data[i]);
+    }
+
+    // expected auto = const int64_t
+    auto* x2_data = x2->data<int64_t>();
+    // e.g get [0, 1, 2, 3, ...]
+    auto x2_lods = x2->lod()[0];
+    auto x1_lods = x1->lod()[0];
+
+    std::unordered_map<int64_t, int64_t> mmap_aux;
+    std::vector<size_t> ins_after_filter;
+    Vector<size_t> out_lods(1, 0);
+    for (size_t i = 0; i < x2_lods.size() - 1; i++) {
+      for (size_t j = x2_lods[i]; j < x2_lods[i + 1]; j++) {
+        if (filter_tag.find(x2_data[j]) != filter_tag.end()) {
+          ins_after_filter.push_back(x2_lods[i]);
+          size_t batch_len = x1_lods[i + 1] - x1_lods[i];
+          mmap_aux[out_lods.back()] = x1_lods[i];
+          out_lods.push_back(out_lods.back() + batch_len);
+          break;
+        }
+      }
+    }
+
+    // set output value
+    // for those whose ins been dropout, set 0 for whole lines.
+    // otherwise, copy whole line
+    // Dim [local fc count, batch size, embedding size]
+    LoDTensor* out = context.Output<LoDTensor>("Out");
+    LoDTensor* map = context.Output<LoDTensor>("Map");
+    // expected auto = const T
+    auto* x1_data = x1->data<T>();
+    // expected auto = T
+    size_t x1_embed_size = x1->dims()[1];
+    out->Resize(framework::make_ddim(
+        {(int64_t)out_lods.back(), (int64_t)x1_embed_size}));
+    map->Resize(framework::make_ddim({(int64_t)ins_after_filter.size(), 3}));
+    auto* out_data = out->mutable_data<T>(context.GetPlace());
+    auto* map_data = map->mutable_data<int64_t>(context.GetPlace());
+
+    Vector<size_t> map_lods;
+    for (size_t i = 0; i < ins_after_filter.size(); i++) {
+      map_data[i * 3] = (int64_t)out_lods[i];
+      map_data[i * 3 + 1] = mmap_aux[map_data[i * 3]];
+      map_data[i * 3 + 2] = out_lods[i + 1] - out_lods[i];
+      map_lods.push_back(i);
+    }
+    map_lods.push_back(ins_after_filter.size());
+    std::vector<Vector<size_t>> map_lod_info;
+    map_lod_info.push_back(map_lods);
+
+    map->set_lod(map_lod_info);
+
+    std::vector<Vector<size_t>> out_lod_info;
+    out_lod_info.push_back(out_lods);
+    out->set_lod(out_lod_info);
+    memset(out_data, 0, out->dims()[0] * out->dims()[1] * sizeof(T));
```

Comment:
why memset here? I think you do memcpy for all elements.

Reply:
thank you, I've deleted this.